go-controller/pkg/network-controller-manager/netattachdefcontroller.go

```diff
@@ -0,0 +1,544 @@
+}
+
+type nadNameInfo struct {
+	nadNames  map[string]bool
+	oc        NetworkController
+	isStarted bool
+}
+
+type nadNetConfInfo struct {
+	util.NetConfInfo
+	netName string
+}
+
+// GetAllNetworkControllers returns a snapshot of all managed nad associated network controllers.
+// Caller needs to note that there are no guarantees the return results reflect the real time
+// condition. There maybe more controllers being added, and returned controllers nay be deleted
+func (cnm *netAttachDefManager) GetAllNetworkControllers() []NetworkController {
+	allNetworkNames := cnm.perNetworkNadNameInfo.GetKeys()
+	allNetworkControllers := make([]NetworkController, 0, len(allNetworkNames))
+	for _, netName := range allNetworkNames {
+		cnm.perNetworkNadNameInfo.LockKey(netName)
+		nni, ok := cnm.perNetworkNadNameInfo.Load(netName)
+		if ok {
+			allNetworkControllers = append(allNetworkControllers, nni.oc)
+		}
+		cnm.perNetworkNadNameInfo.UnlockKey(netName)
+	}
+	return allNetworkControllers
+}
+
+// AddNetAttachDef adds the given nad to the associated controller. It creates the controller if this
+// is the first nad of the network
+// note that for errors that are not retriable (configuration error etc.), just log the error and return nil
+func (cnm *netAttachDefManager) AddNetAttachDef(ncm NetworkControllerManager,
+	netattachdef *nettypes.NetworkAttachmentDefinition, doStart bool) error {
+	var netConfInfo util.NetConfInfo
+	var nInfo util.NetInfo
+	var err error
+
+	netAttachDefName := util.GetNadName(netattachdef.Namespace, netattachdef.Name)
+	klog.Infof("Add net-attach-def %s", netAttachDefName)
+
+	nInfo, netConfInfo, err = util.ParseNADInfo(netattachdef)
+	netName := ""
+	if err == nil {
+		netName = nInfo.GetNetworkName()
+	}
+
+	return cnm.perNadNetConfInfo.DoWithLock(netAttachDefName, func(nadName string) error {
+		nadNci, loaded := cnm.perNadNetConfInfo.LoadOrStore(nadName, &nadNetConfInfo{
+			NetConfInfo: netConfInfo,
+			netName:     netName,
+		})
+		if !loaded {
+			// first time to process this nad
+			if err != nil {
+				// invalid nad, nothing to do
+				klog.Warningf("net-attach-def %s is first seen and is invalid: %v", nadName, err)
+				cnm.perNadNetConfInfo.Delete(nadName)
+				return nil
+			}
+			klog.V(5).Infof("net-attach-def %s network %s first seen", nadName, netName)
+			err = cnm.addNadToController(ncm, nadName, nInfo, netConfInfo, doStart)
+			if err != nil {
+				klog.Errorf("Failed to add net-attach-def %s to network %s: %v", nadName, netName, err)
+				cnm.perNadNetConfInfo.Delete(nadName)
+				return err
+			}
+		} else {
+			klog.V(5).Infof("net-attach-def %s network %s already exists", nadName, netName)
+			nadUpdated := false
+			if err != nil {
+				nadUpdated = true
+			} else if nadNci.netName != netName {
+				// netconf network name changed
+				klog.V(5).Infof("net-attach-def %s network name %s has changed", netName, nadNci.netName)
+				nadUpdated = true
```

Comment:
I'm a bit confused about this block; what does it represent for an existing attachment to have a new net name ? 

Why is it important to isolate from the case below ? i.e. `else if !nadNci.CompareNetConf(netConfInfo)`

Reply:
nothing but the details of the change that is part of the log